### ISSUE TITLE
[macOS] Adopt Base Internationalization

### ISSUE THREAD

**Issue by nevack** [transmission/transmission#3405]
[Related Apple Documentation in Archive](https://developer.apple.com/library/archive/documentation/MacOSX/Conceptual/BPInternational/InternationalizingYourUserInterface/InternationalizingYourUserInterface.html)

Basically, instead of having multiple XIBs for every supported locale, we could use single XIB per Window or View and provide all translations in `*.strings` files.

Related PR https://github.com/transmission/transmission/pull/3039

**Comment by sweetppro** [comment]
The Preferences Window looks like the most time intensive window to implement Autolayout. All the other views are basically done, or not that difficult to implement

**Comment by Oleg-Chashko** [comment]
@nevack 
If you give me a specific purpose for what to check, I'll be happy to help you check it.

**Comment by Oleg-Chashko** [comment]
@nevack 
And indicate in which issue to put the test reports.

**Comment by nevack** [comment]
> @nevack 
> 
> If you give me a specific purpose for what to check, I'll be happy to help you check it.

We have two connected issues here.
We want to change the way UI loads translations.
Instead of having each language it's own layout, we will use only single one.
But to do this properly, we need to upgrade our layouts from fixed to sizing to Auto Layout with Constraints.

So we need to check that every window has no sizing or clipping issues with all locales.

**Comment by nevack** [comment]
> @nevack 
> 
> And indicate in which issue to put the test reports.

This issue is fine, when we're done here and some new pesky bug found, you are free to fill in a new one.

**Comment by Oleg-Chashko** [comment]
@nevack 

Will you let me know when the PR is up so I can start testing?

**Comment by nevack** [comment]
> @nevack 
> 
> 
> 
> Will you let me know when the PR is up so I can start testing?

Of course, for now we need OK from @ckerr to start.

**Comment by sweetppro** [comment]
I’m out for the day,
But can get something prepared tonight unless someone else wants to get the ball rolling quicker.

I’ll start piecemeal with the simplest xibs.

As I mentioned earlier, the Preferences window will take the most effort. All the others will take little to no effort.

**Comment by ckerr** [comment]
Well this is cool. I had assumed it would be a significant amount of work. This is what I get for making assumptions about the macOS client.

Yes if all of you are in agreement and think it's reasonably easy, then let's do it.

**Comment by nevack** [comment]
@ckerr I think it would be safer to create new branch for this specific issue.
Like, `macos-base-internalization`.
This way trunk would stay green, as soon we are ready, then we will create PR from this branch to `main`.

**Comment by nevack** [comment]
> > @ckerr I think it would be safer to create new branch for this specific issue. Like, `macos-base-internalization`. This way trunk would stay green, as soon we are ready, then we will create PR from this branch to `main`.
> 
> I literally have no idea what is involved in the base i18n process, so whatever you all collectively want to do is fine with me

Ok, I request from maintainers to create `macos-base-internalization` branch from `main`, current PR will be retargeted into it. :wink:

**Comment by ckerr** [comment]
Done. https://github.com/transmission/transmission/compare/macos-base-internalization?expand=1

**Comment by killemov** [comment]
> Done. https://github.com/transmission/transmission/compare/macos-base-internalization?expand=1

I had to read it a couple of times myself but a simple copy-paste action/mistake has now made "internalization" (i15n) an eternal feature branch. Or .. What is being internalized here?

But the actual question is: Why is this done specifically for the MacOS client? AFAIK ALL (>99% ?) strings are common in all versions. Or is the MacOS version just the most convenient to start with?

**Comment by killemov** [comment]
> It's macOS build specific

Not ok.
> newer way of importing translations.

Possibly a good development?

... IF the resulting string files are then also used for all other versions. I have little knowledge of the source but I have to assume these strings are loaded at runtime. If not then that is also not good. My point being is you'd eventually want a single source of translated strings for all versions to simplify both the code base and future translations.

**Comment by nevack** [comment]
> Not ok.

Please, elaborate.

> ... IF the resulting string files are then also used for all other versions.

macOS, Qt and GTK clients are different as they use different format of translation files.
All strings come from transifex, but they are converted for specific ToolKit.

Also, not all string translations present on all platforms, as some features are exclusive to some platform.

I don't understand what you really mean here.

**Comment by ckerr** [comment]
> macOS, Qt and GTK clients are different as they use different format of translation files.
All strings come from transifex, but they are converted for specific ToolKit.

Slightly off-topic for this thread, but it sounds like this _might_ be useful for planning longer term?

The overlap of translatable strings between transmission-gtk and transmission-qt is probably >90%. Long-term IMO it would be worthwhile to have a single set of translatable strings and feed it into those clients, so that translators don't have to duplicate their work. Accomplishing this is nontrivial but it is a longer-term goal.

However I _don't_ know how much overlap there is with the macOS client

**Comment by killemov** [comment]
> > Not ok.
> 
> Please, elaborate.

Ah, doing something for a specific platform moves the code base out of center for all platforms. If the Mac-client is so very different from Linux or Windows (and IMHO Web as well) that it warrants specific parallel development then THAT is not ok. A logical exception could be a phone-specific interface because of very different interface requirements.

> 
> > ... IF the resulting string files are then also used for all other versions.
> 
> macOS, Qt and GTK clients are different as they use different format of translation files. All strings come from transifex, but they are converted for specific ToolKit.

AFAIK Transifex supports many formats but does not designate one format to rule them all. If the maintenance of translations is kept within a single file per language for all platforms then my point is moot, as that is or should be the goal. 

> Also, not all string translations present on all platforms, as some features are exclusive to some platform.

I see no problems with having more strings in translation files than are used on any specific platform. (Maybe a small one for cross-platform testing.) I see problems with having duplicate strings in translation files per platform.

> I don't understand what you really mean here.

To summarize: It's recommended to keep all translation strings for all platforms in a single file per language. And, if need be or more convenient, the build process could transform these files in the required format. (But not doing that and using the source translation files is better.)

**Comment by nevack** [comment]
> Ah, doing something for a specific platform moves the code base out of center for all platforms.

macOS translations string are already separated, so GTK and Qt.

> If the Mac-client is so very different from Linux or Windows (and IMHO Web as well) that it warrants specific parallel development then THAT is not ok.

GTK, Qt and Web clients are cross-platform, macOS client is native.

> I see problems with having duplicate strings in translation files per platform.
> If the maintenance of translations is kept within a single file per language for all platforms then my point is moot, as that is or should be the goal.
> To summarize: It's recommended to keep all translation strings for all platforms in a single file per language. And, if need be or more convenient, the build process could transform these files in the required format. (But not doing that and using the source translation files is better.)

Sorry, but you are being off-topic.

This issue is about converting existing macOS-specific translations to another format.
Right now every locale had its own design/layout file called XIB.
We want to have single XIB per window and import translation from per-locale `.strings` file.
It's way too easier to manage, as layout can be changed once for all languages.

This issue is not about diverging macOS translations from other clients, or about single source of translation keys.

I have tried to point this out, but not luck.

If you want to discuss the whole translation process - feel free to file another issue.

**Comment by killemov** [comment]
> > Ah, doing something for a specific platform moves the code base out of center for all platforms.
> 
> macOS translations string are already separated, so GTK and Qt.

So my suspicions of the wheel being reinvented for every platform are confirmed. I DO get that this is not your problem at hand, but with the C++ overhaul in progress, rethinking radical integration might be something to consider.

> > If the Mac-client is so very different from Linux or Windows (and IMHO Web as well) that it warrants specific parallel development then THAT is not ok.
> 
> GTK, Qt and Web clients are cross-platform, macOS client is native.

I think we have a different concept of "native". The GTK, Qt, MacOS and Windows clients each implement the interface natively with Web and CLI being the only exceptions.

> > I see problems with having duplicate strings in translation files per platform.
> > If the maintenance of translations is kept within a single file per language for all platforms then my point is moot, as that is or should be the goal.
> > To summarize: It's recommended to keep all translation strings for all platforms in a single file per language. And, if need be or more convenient, the build process could transform these files in the required format. (But not doing that and using the source translation files is better.)
> 
> Sorry, but you are being off-topic.

With the MacOS client being as separate as you point out it is, I get that.
> 
> This issue is about converting existing macOS-specific translations to another format. Right now every locale had its own design/layout file called XIB. We want to have single XIB per window and import translation from per-locale `.strings` file. It's way  easier to manage, as layout can be changed once for all languages.

I know layout puts constraints on rendered string length (W != I), and vice versa, but does this need to be more complicated than accommodating the wordiest language so everything else will fit anyway? The way you explain what "we" want appears to be both  flexible and maintenance intensive. And are shared strings really shared or duplicated for every window? (If automated then also no problem.)
> 
> This issue is not about diverging macOS translations from other clients, or about single source of translation keys.

It appears that damage(?) was done some time ago anyway. :smile: 
 
> I have tried to point this out, but not luck.
> 
> If you want to discuss the whole translation process - feel free to file another issue.

I think @ckerr suggested a path in the right direction.

**Comment by nevack** [comment]
> So my suspicions of the wheel being reinvented for every platform are confirmed. I DO get that this is not your problem at hand, but with the C++ overhaul in progress, rethinking radical integration might be something to consider.

But why do you still keeping this conversation in an somewhat-related issue?
It's better to create a separate one for the history.
All your concerns are important, but it's off the scope right here.

> I think we have a different concept of "native". The GTK, Qt, MacOS and Windows clients each implement the interface natively with Web and CLI being the only exceptions.

macOS client uses Cocoa API, it is the native toolkit from Apple.
I know GTK and Qt compiles to WinAPI or Xlib/Wayland drawing calls, the are in some way native in the contrary to something like Electron, yes.
That's not my point.
My point is macOS client is totally separate and can have some specific features.

> With the MacOS client being as separate as you point out it is, I get that.

Yes!

> I know layout puts constraints on rendered string length (W != I), and vice versa, but does this need to be more complicated than accommodating the wordiest language so everything else will fit anyway? The way you explain what "we" want appears to be both flexible and maintenance intensive. And are shared strings really shared or duplicated for every window? (If automated then also no problem.)

Yes, and that's why we are converting this layouts from being fixed width to AutoLayout.
If you open [the link from issue description](https://developer.apple.com/library/archive/documentation/MacOSX/Conceptual/BPInternational/InternationalizingYourUserInterface/InternationalizingYourUserInterface.html) you'll see there's even double-width Pseudolocalization to help us test it.

> It appears that damage(?) was done some time ago anyway. 😄

I have joined the project recently, so yes, I have nothing to do about how translation process works.
And I don't think any damage have been done, it's easy to judge some choices in project such mature as Transmission is.
Yes, there can be a lot of room for improvements, you can highlight some of them in *a separate issue*.

I just don't understand, why I feel like I need to defend something should be done with macOS just because totally separate problem is not being solved simultaneously.

**Comment by ckerr** [comment]
I kind of agree with both of you -- combining the strings all together would be a good thing, but (unless someone steps up to volunteer on the work) it's not going to happen soon and shouldn't block Base I18n.

Probably would be a good idea to open up a separate tracking issue e.g. if someone wants to volunteer to do the work.

**Comment by ckerr** [comment]
Is this issue finished, e.g. should I close this as fixed?

**Comment by ckerr** [comment]
@transmission/contributors is this issue finished, e.g. should I close this issue as fixed?

It's milestoned for 4.0.0-beta.1

**Comment by Pentaphon** [comment]
@nevack if the only blocker for 4.0.0-beta.1 let us know

**Comment by nevack** [comment]
> @nevack if this is the only blocker for 4.0.0-beta.1 let us know

I think we can close this now.
Thanks to @sweetppro (dev) and @Oleg-Chashko (QA) 🎉

**Comment by Oleg-Chashko** [comment]
I'm glad I could help.
